vertical-pod-autoscaler/pkg/updater/priority/scaling_direction_pod_eviction_admission.go

```diff
@@ -0,0 +1,104 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package priority
+
+import (
+	apiv1 "k8s.io/api/core/v1"
+	vpa_types "k8s.io/autoscaler/vertical-pod-autoscaler/pkg/apis/autoscaling.k8s.io/v1"
+	vpa_utils "k8s.io/autoscaler/vertical-pod-autoscaler/pkg/utils/vpa"
+)
+
+// NewScalingDirectionPodEvictionAdmission creates a PodEvictionAdmission object.
+// It admits Pods for eviction based on the scaling direction, i.e. if a resource is scaled up (recommendation > requests) or scaled down (recommendation < requests).
+// It also supports >= and <= relations.
+func NewScalingDirectionPodEvictionAdmission() PodEvictionAdmission {
+	return &scalingDirectionPodEvictionAdmission{}
+}
+
+type scalingDirectionPodEvictionAdmission struct {
+	EvictionRequirements map[*apiv1.Pod][]*vpa_types.EvictionRequirement
+}
+
+// Admit denies Pod eviction, if an EvictionRequirement exists for at least one Container which evaluates to 'false'.
+// If no EvictionRequirement exists or no Recommendation is present yet, the Pod is admitted for eviction.
+func (s *scalingDirectionPodEvictionAdmission) Admit(pod *apiv1.Pod, resources *vpa_types.RecommendedPodResources) bool {
+	podEvictionRequirements, found := s.EvictionRequirements[pod]
+	if !found {
+		return true
+	}
+	for _, container := range pod.Spec.Containers {
+		recommendedResources := vpa_utils.GetRecommendationForContainer(container.Name, resources)
+		if recommendedResources == nil {
```

Comment:
This seems incorrect. This will admit pod if at least one container:
- doesn't have a recommendation or
- has a recommendation in the "correct" direction

I think this should return false if it finds at leas 1 container that has recommendation going in the wrong direction, default to true
?

Reply:
Description of the function[1] and AEP[2] say that all `EvictionRequirement`s must be met to allow eviction. 

If that's not what you need then we can revisit the KEP (at least rename and have the description match implementation; I didn't think to much about if allowing eviction if at least one container matches is better than requiring all of them match restriction).

[1]
> Admit denies Pod eviction, if an EvictionRequirement exists for at least one Container which evaluates to 'false'.

[2]
[permalink](https://github.com/kubernetes/autoscaler/blob/d3ec0c4d049084bc5bb266a19448df0e82e3056a/vertical-pod-autoscaler/enhancements/4831-control-eviction-behavior/README.md?plain=1#L24):

> A single `EvictionRequirement` defines a condition which must be `true` to allow eviction for the corresponding `Pod`. When multiple `EvictionRequirements` are specified for a `Pod`, all of them must evaluate to `true` to allow eviction.

Reply:
I think what the AEP mentions still holds true:
> A single EvictionRequirement defines a condition which must be true to allow eviction for the corresponding Pod. When multiple EvictionRequirements are specified for a Pod, all of them must evaluate to true to allow eviction.

In [`allowEvictionForResourceAndScalingDirection`](https://github.com/kubernetes/autoscaler/blob/dcec527d41e1604c64609ea6640833242db9f85a/vertical-pod-autoscaler/pkg/updater/priority/scaling_direction_pod_eviction_admission.go#L71) we iterate over all `EvictionRequirements` and if _one_ of the Requirements returns `false`, we return `false` for the entire Pod.

The critical point is _when_ an `EvictionRequirement` is `true` for a `Pod` based on the individual evaluations for the `Containers`.

I realize that the current comment is incorrect, I think I moved this one back and forth a few times
> Admit denies Pod eviction, if an EvictionRequirement exists for at least one Container which evaluates to 'false'.

This probably should be 
> Admit admits a Pod for eviction in one of three cases 
>  * an EvictionRequirement exists which is evaluated to 'true' for at least one Container in this Pod
>  * no Recommendation exists for at least one Container in this Pod
>  * no EvictionRequirement exists for this Pod

Does this make sense? Do we need a clarification in the AEP as well regarding what it means that an `EvictionRequirement` is true for a `Pod`?

Reply:
Thanks for the explanation. I'll think about this a bit, I hope I'll be back with a proper reply on Monday

Reply:
My understanding of the AEP:

- If there is only 1 eviction requirement, on 1 resource we don't evict unless one resource changes in the specified direction
- If there is only 1 eviction requirement, on both resources we don't evict unless at least one resource changed 
- If there are 2 eviction requirements, on both resources separately we don't evict unless both resources change in their specified directions

This code treats separate `EvictionRequirement`s  for CPU the way KEP describes putting them in a single `EvictionRequirement` (one of them changes in the specified direction, not necessarily both). This seems wrong to me

Also I think we should clarify in the AEP that "A single EvictionRequirement defines a condition which must be true to allow eviction for the corresponding Pod" means that at least one container has a change in the specified direction. This makes sense but got me confused for a while

Reply:
Good catch, thanks! I think there was a test case missing that covered the scenario of a single `EvictionRequirement` with multiple resources, where only one of them was matched and the Pod was still admitted.
I added the case and fixed the code to make it green.

The other scenario ensuring that multiple `EvictionRequirements` are connected with `AND` was already in the test cases and still works.

I've also restructured the code a bit to match the comments and early exists for the cases in which we admit a Pod, hope this makes more sense now?

Reply:
Opened PR https://github.com/kubernetes/autoscaler/pull/5989 to address this part of the comment 

> Also I think we should clarify in the AEP that "A single EvictionRequirement defines a condition which must be true to allow eviction for the corresponding Pod" means that at least one container has a change in the specified direction.